Drop storage replica when dropping database replica with a zookeeper path

tuanpach opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
- Rename InterpreterSystemQuery::dropReplicaImpl to InterpreterSystemQuery::dropStorageReplica
- In InterpreterSystemQuery::dropDatabaseReplica:
    - When dropping with database or drop the whole replica: it also drops replica for each table of the database
         - If 'WITH TABLES' is provided, drop replica for each storage
         - Otherwise, the logic is unchanged, only call DatabaseReplicated::dropReplica on the databases
    - When dropping a database replica with the keeper path:
        - If 'WITH TABLES' is provided:
            - Restore the database as Atomic
            - Restore RMT tables from statement in Keeper
            - Drop the database (restored tables are also dropped)
        - Otherwise, only call DatabaseReplicated::dropReplica on the provided keeper path


Closes https://github.com/ClickHouse/clickhouse-private/issues/35885

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

evillique commented on tests/integration/test_drop_database_replica/test.py:
```diff
@@ -0,0 +1,215 @@
+    )
+
+    node4.query(
+        "CREATE TABLE db.rmt (x INT) ENGINE=ReplicatedMergeTree ORDER BY x",
```
I would also check whether `SYSTEM DROP REPLICA` would remove the ZK nodes of the replica and the table.

evillique commented on src/Interpreters/InterpreterSystemQuery.cpp:
```diff
@@ -1219,26 +1229,193 @@ void InterpreterSystemQuery::dropReplica(ASTSystemQuery & query)
+                /*zookeeper_path_=*/zookeeper_path,
+                /*node_name=*/table_name,
+                /*query=*/create_query_string);
+            auto create_query_context = Context::createCopy(getContext());
```
We need to enable all the experimental settings, and a lot of other things, similar to `recoverLostReplica` in `DatabaseReplicated`:

https://github.com/ClickHouse/ClickHouse/blob/ced5264967018bc2fdfdcc5edf6cbc8be7ae89c2/src/Databases/DatabaseReplicated.cpp#L1369-L1395

I would also probably to it with `LoadingStrictnessLevel::FORCE_RESTORE`.

evillique commented on tests/integration/test_drop_database_replica/test.py:
```diff
@@ -0,0 +1,215 @@
+        },
+    )
+    node4.query(f"SYSTEM DROP REPLICA 'dummy' FROM DATABASE db")
+    node4.query(f"SYSTEM DROP REPLICA 'dummy'")
```
The final thing I would check is a Refreshable Materialized View; it behaves differently based on the database it is being created in, and if I remember correctly, it may refuse to be created in Atomic DB with some parameters.

evillique commented on src/Interpreters/InterpreterSystemQuery.cpp:
```diff
@@ -1219,26 +1229,193 @@ void InterpreterSystemQuery::dropReplica(ASTSystemQuery & query)
+
+        table_name_to_metadata.emplace(unescapeForFileName(escaped_table_names[i]), std::move(res.data));
+    }
+    String create_query = fmt::format("CREATE DATABASE `{}` ENGINE=Atomic", restoring_database_name);
```
> Yes. Should we add logic to clean up this database when the instance starts up?

Well, at least that is definitely needed, but I don't particularly like making a persistent DB with real tables at all.

> Do you think we should only restore SharedMergeTree tables, or storages containing SharedMergeTree inner table?

Well, the current option is ok and universal, but not optimal in terms of the things I think RMT will start doing as soon as we create it. For SMT there is no other way than to create the table.

The ideal way I see it is that we go through table definitions, call dropReplica for RMT tables, create SMT tables (without even creating a database, call drop replica, and then drop this table), and skip all the other engines. The only thing that needs investigation is whether it can be broken by not having some tables and the dependencies somehow breaking. This way this would execute as fast as it can and require the minimum amount of resources. But I don't know whether this level of optimization is needed for a SYSTEM query.

tavplubix replied:
Can we just create the database object in memory (without executing the CREATE query and adding the database to DatabaseCatalog) and call `drop()` or something?

XuJia0210 commented on src/Interpreters/InterpreterSystemQuery.cpp:
```diff
@@ -1303,6 +1482,40 @@ void InterpreterSystemQuery::dropDatabaseReplica(ASTSystemQuery & query)
+            auto database = restoreDatabaseFromKeeperPath(query.replica_zk_path, restoring_database_name);
+            if (database)
+            {
+                dropStorageReplicasFromDatabase(query.replica, database);
```
We don't need to call `dropStorageReplicasFromDatabase` here
Since we will run the query `DROP DATABASE`

tuanpach committed:
- Rename InterpreterSystemQuery::dropReplicaImpl to InterpreterSystemQuery::dropStorageReplica
- In InterpreterSystemQuery::dropDatabaseReplica:
    - When dropping with database or drop the whole replica: it also drops replica for each table of the database
    - When dropping a database replica with the keeper path:
        - If RESTORE_AND_DROP_DATABASE is provided:
            - Restore the database as Atomic
            - Restore tables from statement in Keeper, skip Materialized View
            - Drop the database (restored tables are also dropped)
        - Otherwise, only call DatabaseReplicated::dropReplica on the provided keeper path

tuanpach committed:
- Use 'WITH TABLES' instead of RESTORE_AND_DROP_DATABASE
- When restoring databases for table dropping, skip restoring non-ReplicatedMergeTree tables
- After loading databases, clean up the restoring databases for table dropping

tuanpach committed:
Merge branch 'master' into drop-storage-replicas-when-dropping-database-replica

tuanpach committed:
Update InterpreterSystemQuery.cpp

Fix compilation by removing a duplicate include

tuanpach committed:
Merge branch 'master' into drop-storage-replicas-when-dropping-database-replica
